File: packages/amplify_auth_cognito/ios/Classes/FlutterSignUpRequest.swift

/*
 * Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * A copy of the License is located at
 *
 *  http://aws.amazon.com/apache2.0
 *
 * or in the "license" file accompanying this file. This file is distributed
 * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import Foundation
import Amplify
import AmplifyPlugins
import amplify_core

struct FlutterSignUpRequest {
    var username: String
    var password: String
    var options: AuthSignUpRequest.Options?
    init(dict: NSMutableDictionary){
        self.username = dict["username"] as! String
        self.password = dict["password"] as! String
        self.options = createOptions(options: dict["options"] as? Dictionary<String, Any>)
    }
    
    private func createOptions(options: Dictionary<String, Any>?) -> AuthSignUpRequest.Options? {
        var userAttributes: [AuthUserAttribute]? = nil
        var pluginOptions: AWSAuthSignUpOptions? = nil
        if (options?["userAttributes"] is Dictionary<String, String>) {
            let rawAttributes = options!["userAttributes"] as! Dictionary<String, String>
            userAttributes = createAuthUserAttributes(rawAttributes: rawAttributes)
        }
        
        if (options?["clientMetadata"] is Dictionary<String, String> || options?["validationData"] is Dictionary<String, String>) {
            pluginOptions = AWSAuthSignUpOptions(
                validationData: options?["validationData"] as? [String: String],
                metadata: options?["clientMetadata"] as? [String : String]
            )
        }
        return AuthSignUpRequest.Options(userAttributes: userAttributes, pluginOptions: pluginOptions)
    }
    
    private func createAuthUserAttributes(rawAttributes: Dictionary<String, String>) -> [AuthUserAttribute] {
        return rawAttributes.map { attr in
            createAuthUserAttribute(key: attr.key, value: attr.value)
        }
    }
    
    static func validate(dict: NSMutableDictionary) throws {
        let validationErrorMessage = "SignUp Request malformed."
        if (dict["password"] == nil) {
            throw InvalidRequestError.auth(comment: validationErrorMessage,
                                           suggestion: String(format: ErrorMessages.missingAttribute, "password "))
        }
    }
}

File: packages/auth/amplify_auth_cognito_android/android/src/main/kotlin/com/amazonaws/amplify/amplify_auth_cognito/types/FlutterSignUpRequest.kt
/*
 * Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * A copy of the License is located at
 *
 *  http://aws.amazon.com/apache2.0
 *
 * or in the "license" file accompanying this file. This file is distributed
 * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

@file:Suppress("UNCHECKED_CAST")

package com.amazonaws.amplify.amplify_auth_cognito.types

import com.amazonaws.amplify.amplify_auth_cognito.utils.createAuthUserAttribute
import com.amazonaws.amplify.amplify_core.exception.ExceptionMessages
import com.amazonaws.amplify.amplify_core.exception.InvalidRequestException
import com.amplifyframework.auth.AuthUserAttribute
import com.amplifyframework.auth.cognito.options.AWSCognitoAuthSignUpOptions

data class FlutterSignUpRequest(val map: HashMap<String, *>) {
    val username: String = map["username"] as String
    val password: String = map["password"] as String
    val options: AWSCognitoAuthSignUpOptions = formatOptions(map["options"] as HashMap<String, *>?)

    private fun formatOptions(rawOptions: HashMap<String, *>?): AWSCognitoAuthSignUpOptions {
        val optionsBuilder = AWSCognitoAuthSignUpOptions.builder()
        if (rawOptions != null) {
            val attributeData = rawOptions["userAttributes"] as? MutableMap<String, String>
            val clientMetadata = rawOptions["clientMetadata"] as? MutableMap<String, String>
            val validationData = rawOptions["validationData"] as? MutableMap<String, String>

            if (attributeData is MutableMap<String, String>) {
                val authUserAttributes: List<AuthUserAttribute> = attributeData.map { (key, value) ->
                    createAuthUserAttribute(key, value)
                }
                optionsBuilder.userAttributes(authUserAttributes)
            }

            if (validationData is MutableMap<String, String>) {
                optionsBuilder.validationData(validationData)
            }

            if (clientMetadata is MutableMap<String, String>) {
                optionsBuilder.clientMetadata(clientMetadata)
            }
        }
        return optionsBuilder.build()
    }

    companion object {
        private const val validationErrorMessage: String = "SignUp Request malformed."
        fun validate(req: HashMap<String, *>?) {
            if (req == null) {
                throw InvalidRequestException(validationErrorMessage, ExceptionMessages.missingAttribute.format("request map"))
            }
            if (!req.containsKey("password")) {
                throw InvalidRequestException(validationErrorMessage, ExceptionMessages.missingAttribute.format("password"))
            }
        }
    }
}
